src/librbd/plugin/Api.cc

```diff
@@ -26,13 +26,15 @@ void Api<I>::execute_image_metadata_set(
     I *image_ctx, const std::string &key,
     const std::string &value, Context *on_finish) {
   ImageCtx* ictx = util::get_image_ctx(image_ctx);
+  std::unique_lock owner_lock{ictx->owner_lock};
```

Comment:
We don't take the unique_lock on `owner_lock` when doing operations -- why

Reply:
Originally the lock was acquired in ShutdownRequest<I>::send_remove_image_cache_state(), ImageCacheState<I>::write_image_cache_state() and ShutdownRequest<I>::send_remove_image_cache_state(). And it acquired shared_lock. 
I moved all the locking here. And I think it should acquire write lock instead of read lock as these are updated operations, so I change the shared lock to unique lock. is it right?

Reply:
The `owner_lock` is acquired as a read-lock for all operations just to prevent the lock from being released concurrently. The actual changing of state happens under a write lock.

Reply:
> Prevent the lock from being released concurrently

The lock means the lock by exclusive lock, right? 
It doesn't allow that when the operations like metadata_set/remove is in progress while the exclusive lock is being released. Right?

Reply:
Correct -- i.e. just follow the existing code behavior.

Reply:
Thank you. I will close this PR.